File: dliscript.py
import sys
import os
import urllib2
import re
import codecs
import time
import BeautifulSoup

def getbookinfo(baseurl, html):
	soup = BeautifulSoup.BeautifulSoup(html)
	table = soup.find('table', border='2', cellpadding='3', width='80%')
	rows = table.findAll('tr')
	for row in rows[1:]:
		titleandlink = row.contents[1]
		author = row.contents[3].text
		pages = row.contents[5].text
		title = titleandlink.text
		selectionregex = re.compile(r'^\.\./manuscript/.*')
		anchortag = titleandlink.find('a', href=selectionregex)
		link = anchortag.attrMap['href'].rpartition('/')[0]
		if baseurl[-1] == '/':
			link = baseurl + link
		else:
			link = baseurl + '/' + link
		print title
		print author
		print pages
		print link
		yield title, author, pages, link

def downloadpages(link, start, count, targetdestination):
	for i in xrange(0, count):
		pagelink = link + '/' + str(start + i)
		try:
			response = urllib2.urlopen(pagelink)
		except:
			print 'Failed to download %s. Skipping...' % pagelink
			continue
		else:
			html = response.read()
			soup = BeautifulSoup.BeautifulSoup(html)
			anchortag = soup.find('a', id='ctl00_ContentPlaceHolder1_imageLink')
			imagelink = anchortag.attrMap['href']
			imagefilename = imagelink.split('/')[-1]
			print 'Downloading image %s to file %s' % (imagelink, imagefilename)
			with open(os.path.join(targetdestination, imagefilename), 'wb') as imagefile:
				try:
					imageresponse = urllib2.urlopen(imagelink)
					image = imageresponse.read()
					imagefile.write(image)
					imagefile.close()
				except:
					print 'Failed to download %s' % imagelink
		pausetime = 1
		print 'Sleeping for %d second(s) before requesting the next image...' % pausetime
		time.sleep(pausetime)
	
def setupbook(rootfolder, title, author, pages, link):
	currentfolderpath = os.path.join(rootfolder, title)
	# check if the target folder exists.
	if not os.path.exists(os.path.join(rootfolder, title)):
		os.mkdir(os.path.join(rootfolder, title))
	# check if book metadata exists in the target folder.
	metadatafile = os.path.join(currentfolderpath, ".meta")
	if not os.path.exists(metadatafile):
		with codecs.open(metadatafile, 'w', encoding='utf-8') as metafile:
			metafile.write(','.join([title, author, pages, link]))
			metafile.close()
	
	filesalreadydownloaded = 0
	for root, dirs, files in os.walk(currentfolderpath):
		for f in files:
			currentfilepath = os.path.join(root, f)
			filename, ext = os.path.splitext(f)
			if ext == ".JPG":
				filesalreadydownloaded += 1
	return currentfolderpath, filesalreadydownloaded
	
def main():
	rootfolder = sys.argv[1]
	print 'using %s as root folder' % rootfolder

	# make sure the root folder exists.
	if not os.path.isdir(rootfolder):
		os.mkdir(rootfolder)

    baseurl = 'www.new1.dli.ernet.in/cgi-bin/advsearch_db.cgi?listStart=%d&language1=Urdu&perPage=%d'
	# baseurl = "http://makhtota.ksu.edu.sa/"
	# browseurl = baseurl + 'eBrowse/'
	stepSize = 20
	
	for currentindex in xrange(0, 1000, stepSize):
		# currenturl = browseurl + str(currentindex)
		currenturl =  baseurl % (currentIndex, stepSize)
		print 'fetching %s' % currenturl
		currentbookgroupfolder = str(currentindex) + '-' + str(currentindex + stepSize - 1)
		currentbookgrouppath = os.path.join(rootfolder, currentbookgroupfolder)
		if not os.path.exists(currentbookgrouppath):
			os.mkdir(currentbookgrouppath)
		try:
			response = urllib2.urlopen(currenturl)
		except:
			print 'Failed to fetch data for index %d. Skipping...' % currentindex
			continue
		html = response.read()
		for title, author, pages, link in getbookinfo(browseurl, html):
			currentbookpath, filesalreadydownloaded = setupbook(currentbookgrouppath, title, author, pages, link)
			downloadpages(link, filesalreadydownloaded + 1, int(pages) - filesalreadydownloaded, currentbookpath)
		
		pausetime = 10
		print 'Sleeping for %d seconds before downloading new batch...' % pausetime
		time.sleep(pausetime)
		
if __name__ == "__main__":
	main()

File: dliscript.rb
require 'rubygems'
require 'mechanize'
require 'logger'

$domainRoot = "http://www.new1.dli.ernet.in"
$baseUrl = "http://www.new1.dli.ernet.in/cgi-bin/advsearch_db.cgi?listStart=%d&language1=Urdu&perPage=%d"

rootDir = ARGV[0]

Dir.mkdir(rootDir) if not File.directory?(rootDir)
Dir.chdir(rootDir)

$logger = Logger.new('dliscript.rb.log')

class String
  def starts_with?(prefix)
    prefix = prefix.to_s
    self[0, prefix.length] == prefix
  end
end

class BookItem
  attr_accessor :subject
  attr_accessor :year
  attr_accessor :bar_code
  attr_accessor :pages
  attr_accessor :title
  attr_accessor :metadata_link
  attr_accessor :content_link
  attr_accessor :first_page
  attr_accessor :last_page
  attr_accessor :content_path
  
  def initialize(subject=nil, year=nil, bar_code=nil, pages=nil, title=nil, metadata_link=nil)
    @subject = subject
    @year = year
    @bar_code = bar_code
    @pages = pages
    @title = title
    @metadata_link = metadata_link
  end
  
end


def parseLink(link)
  book = BookItem.new  
  tokens = link.split('&')
  tokens.each { |token|
    parts = token.split('=')
    
    book.subject = parts[1].to_s.strip if parts[0] == 'subject1'
    book.year = parts[1].to_s.strip if parts[0] == 'year'
    book.bar_code = parts[1].to_s.strip if parts[0] == 'barcode'
    book.pages = parts[1].to_s.strip.to_i if parts[0] == 'pages'
    book.title = parts[1].to_s.strip if parts[0] == 'title1'
    book.content_path = parts[1].to_s.strip if parts[0] == 'url'
  }

  book.first_page = 1
  book.last_page = book.pages
  book.metadata_link = link
  
  return book  
end

def get_page_name(page)
  page.to_s.rjust(8, '0')
end

def get_page_name_with_extension(page)
  get_page_name(page) + '.tif'
end

def get_page_url(book, page)
  $domainRoot + book.content_path + '/PTIFF/' + get_page_name_with_extension(page)
end

def get_group_dir_name(start, ending)
  start.to_s + '-' + ending.to_s
end

def save_book(book, file)
  file.puts "Title=#{book.title}"
  file.puts "Barcode=#{book.bar_code}"
  file.puts "Pages=#{book.pages.to_s}"
  file.puts "Subject=#{book.subject}"
  file.puts "Year=#{book.year}"
  file.puts "Content-Path=#{book.content_path}"
  file.puts "First-Page=#{book.first_page}"
  file.puts "Last-Page=#{book.last_page}"
  file.puts "Metadata-Link=#{book.metadata_link}"
end

mechanize = Mechanize.new { |agent|
  agent.user_agent_alias = 'Mac Safari'
  agent.log = $logger
}

(0..2000).step(20).each { |index|
  group_dir_name = get_group_dir_name(index, index + 20)
  Dir.mkdir(group_dir_name) if not File.directory?(group_dir_name)
  Dir.chdir(group_dir_name)
  
  items_list_page = mechanize.get($baseUrl % [index, 20])
  books = []
  items_list_page.links.each { |link|
    book = parseLink(link.href) if link.href.starts_with?('metainfo.cgi')
    books.push(book)
  }    
  
  books.each { |book|
    Dir.mkdir book.bar_code if not File.directory? book.bar_code
    Dir.chdir(book.bar_code)
    
    File.open('.meta', 'w+') { |file| save_book(book, file) }
    
    (book.first_page .. book.last_page).each { |page|
      page_url = get_page_url(book, page)
      begin
        $logger.info 'Downloading %s' % page_url
        if not File.exists?(get_page_name_with_extension(page))
          mechanize.get(page_url).save()
        else
          $logger.info 'Skipping download of already existing file %s' % get_page_name_with_extension(page)
        end
      rescue
        $logger.error 'Failed to download %s' % page_url
      end
    }
  }
}
